File: src-tauri/src/lcu/player_sync.rs
//! Pure parser helpers for LCU match-history and mastery responses.
//!
//! These functions are intentionally side-effect-free (no I/O, no DB) so they
//! can be unit-tested without a live League client.
//!
//! ⚠ The LCU match-history endpoint is not part of the official public Riot API.
//! Its format may change between client versions. Parsers must never panic on
//! unexpected shapes — unknown fields are skipped and errors are counted.

/// A single parsed match suitable for inserting into the `matches` DB table.
#[derive(Debug, Clone, PartialEq)]
pub struct ParsedLcuMatch {
    pub match_id: String, // "LCU_{gameId}" — prefixed to avoid clashes with Riot IDs
    pub champion_id: i64,
    pub position: Option<String>,
    pub win: bool,
    pub kills: i64,
    pub deaths: i64,
    pub assists: i64,
    pub duration_secs: i64,
    pub queue_id: i64,
    pub played_at: i64, // Unix timestamp in seconds
}

/// A single parsed mastery entry for the `mastery` DB table.
#[derive(Debug, Clone, PartialEq)]
pub struct ParsedLcuMastery {
    pub champion_id: i64,
    pub level: i64,
    pub points: i64,
    pub last_play_time: Option<i64>,
}

// ── Match history parser ───────────────────────────────────────────────────

/// Parse an LCU `/lol-match-history/v1/products/lol/{puuid}/matches` response.
///
/// Returns only the games where `puuid` was a participant.  Games that cannot
/// be matched back to `puuid` via `participantIdentities` are silently skipped
/// (counted in the caller's `errors` counter instead of panicking).
pub fn parse_lcu_match_history(json: &serde_json::Value, puuid: &str) -> Vec<ParsedLcuMatch> {
    let games = match json["games"]["games"].as_array() {
        Some(g) => g,
        None => return vec![],
    };

    let mut result = Vec::new();

    for game in games {
        if let Some(m) = parse_single_game(game, puuid) {
            result.push(m);
        }
    }

    result
}

fn parse_single_game(game: &serde_json::Value, puuid: &str) -> Option<ParsedLcuMatch> {
    let game_id = game["gameId"].as_i64()?;
    let queue_id = game["queueId"].as_i64().unwrap_or(0);
    let duration_secs = game["gameDuration"].as_i64().unwrap_or(0);
    let played_at = game["gameCreation"].as_i64().unwrap_or(0) / 1000;

    // Find the participantId that belongs to the local player (by puuid)
    let identities = game["participantIdentities"].as_array()?;
    let local_participant_id = identities.iter().find_map(|id| {
        let player_puuid = id["player"]["puuid"].as_str()?;
        if player_puuid == puuid {
            id["participantId"].as_i64()
        } else {
            None
        }
    })?;

    // Find the participant data for that participantId
    let participants = game["participants"].as_array()?;
    let p = participants
        .iter()
        .find(|p| p["participantId"].as_i64() == Some(local_participant_id))?;

    let champion_id = p["championId"].as_i64()?;
    let win = p["stats"]["win"].as_bool().unwrap_or(false);
    let kills = p["stats"]["kills"].as_i64().unwrap_or(0);
    let deaths = p["stats"]["deaths"].as_i64().unwrap_or(0);
    let assists = p["stats"]["assists"].as_i64().unwrap_or(0);

    // Position derived from timeline.lane (e.g., "MIDDLE", "BOTTOM", "TOP", "JUNGLE")
    let position = p["timeline"]["lane"]
        .as_str()
        .filter(|s| !s.is_empty() && *s != "NONE")
        .map(str::to_string);

    Some(ParsedLcuMatch {
        match_id: format!("LCU_{game_id}"),
        champion_id,
        position,
        win,
        kills,
        deaths,
        assists,
        duration_secs,
        queue_id,
        played_at,
    })
}

// ── Mastery parser ─────────────────────────────────────────────────────────

/// Parse an LCU `/lol-champion-mastery/v1/local-player/champion-mastery` response.
///
/// Returns all entries where `championId` is parseable.
pub fn parse_lcu_mastery(json: &serde_json::Value) -> Vec<ParsedLcuMastery> {
    let entries = match json.as_array() {
        Some(a) => a,
        None => return vec![],
    };

    entries
        .iter()
        .filter_map(|e| {
            let champion_id = e["championId"].as_i64()?;
            let level = e["championLevel"].as_i64().unwrap_or(0);
            let points = e["championPoints"].as_i64().unwrap_or(0);
            let last_play_time = e["lastPlayTime"].as_i64();
            Some(ParsedLcuMastery {
                champion_id,
                level,
                points,
                last_play_time,
            })
        })
        .collect()
}

// ── Summoner name parsing ──────────────────────────────────────────────────

/// Extract (game_name, tag_line) from an LCU current-summoner response.
///
/// Priority:
/// 1. `gameName` + `tagLine` fields if both present
/// 2. Parse `displayName` as "Name#TAG"
/// 3. Use `displayName` as game_name with empty tag_line
pub fn parse_lcu_summoner_name(summoner: &serde_json::Value) -> (String, String) {
    let game_name = summoner["gameName"]
        .as_str()
        .unwrap_or("")
        .trim()
        .to_string();
    let tag_line = summoner["tagLine"]
        .as_str()
        .unwrap_or("")
        .trim()
        .to_string();

    if !game_name.is_empty() {
        return (game_name, tag_line);
    }

    // Fallback: parse displayName
    let display = summoner["displayName"]
        .as_str()
        .unwrap_or("Summoner")
        .trim()
        .to_string();

    if let Some(idx) = display.rfind('#') {
        let name = display[..idx].to_string();
        let tag = display[idx + 1..].to_string();
        if !name.is_empty() {
            return (name, tag);
        }
    }

    (display, String::new())
}

// ── Unit tests ─────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    const MATCH_HISTORY_FIXTURE: &str = include_str!("../../tests/fixtures/lcu_match_history.json");
    const MASTERY_FIXTURE: &str = include_str!("../../tests/fixtures/lcu_mastery.json");

    const LOCAL_PUUID: &str = "test-puuid-local-player";

    fn match_history() -> serde_json::Value {
        serde_json::from_str(MATCH_HISTORY_FIXTURE).expect("fixture parse failed")
    }

    fn mastery() -> serde_json::Value {
        serde_json::from_str(MASTERY_FIXTURE).expect("fixture parse failed")
    }

    #[test]
    fn parse_lcu_match_history_finds_local_player() {
        let json = match_history();
        let matches = parse_lcu_match_history(&json, LOCAL_PUUID);
        assert_eq!(matches.len(), 2, "should find 2 games for local player");
        assert_eq!(matches[0].match_id, "LCU_9876543210");
        assert_eq!(matches[1].match_id, "LCU_9876543211");
    }

    #[test]
    fn parse_lcu_match_history_extracts_stats() {
        let json = match_history();
        let matches = parse_lcu_match_history(&json, LOCAL_PUUID);
        let first = &matches[0];
        assert_eq!(first.champion_id, 238); // Zed
        assert!(first.win);
        assert_eq!(first.kills, 8);
        assert_eq!(first.deaths, 2);
        assert_eq!(first.assists, 4);
        assert_eq!(first.duration_secs, 1823);
        assert_eq!(first.queue_id, 420);
    }

    #[test]
    fn parse_lcu_match_history_position_from_lane() {
        let json = match_history();
        let matches = parse_lcu_match_history(&json, LOCAL_PUUID);
        assert_eq!(matches[0].position.as_deref(), Some("MIDDLE"));
        assert_eq!(matches[1].position.as_deref(), Some("MIDDLE"));
    }

    #[test]
    fn parse_lcu_mastery_returns_all_entries() {
        let json = mastery();
        let entries = parse_lcu_mastery(&json);
        assert_eq!(entries.len(), 3);
    }

    #[test]
    fn parse_lcu_mastery_extracts_level_and_points() {
        let json = mastery();
        let entries = parse_lcu_mastery(&json);
        assert_eq!(entries[0].champion_id, 238);
        assert_eq!(entries[0].level, 7);
        assert_eq!(entries[0].points, 180_000);
        assert_eq!(entries[0].last_play_time, Some(1_700_000_000));
    }

    #[test]
    fn parse_lcu_summoner_name_gamename_tagline() {
        let v = serde_json::json!({ "gameName": "Berkilic", "tagLine": "TR1" });
        let (name, tag) = parse_lcu_summoner_name(&v);
        assert_eq!(name, "Berkilic");
        assert_eq!(tag, "TR1");
    }

    #[test]
    fn parse_lcu_summoner_name_display_name_with_hash() {
        let v = serde_json::json!({ "displayName": "Berkilic#TR1" });
        let (name, tag) = parse_lcu_summoner_name(&v);
        assert_eq!(name, "Berkilic");
        assert_eq!(tag, "TR1");
    }

    #[test]
    fn parse_lcu_summoner_name_display_name_no_hash() {
        let v = serde_json::json!({ "displayName": "JustAName" });
        let (name, tag) = parse_lcu_summoner_name(&v);
        assert_eq!(name, "JustAName");
        assert_eq!(tag, "");
    }

    #[test]
    fn parse_lcu_match_history_skips_other_player_games() {
        let json = match_history();
        // Player who appears only in game 1, not in game 2
        let matches = parse_lcu_match_history(&json, "other-player-puuid");
        assert_eq!(matches.len(), 1, "only one game for other player");
        assert_eq!(matches[0].match_id, "LCU_9876543210");
    }

    #[test]
    fn parse_lcu_match_history_returns_empty_for_unknown_puuid() {
        let json = match_history();
        let matches = parse_lcu_match_history(&json, "nonexistent-puuid");
        assert!(matches.is_empty());
    }
}

File: desktop/src/main/lcu/player-sync.ts
// Pure parsers for LCU match-history / mastery / current-summoner responses +
// the champion-pool counter — TS port of src-tauri/src/lcu/player_sync.rs and
// src-tauri/src/lcu/champ_pool.rs (`compute_champion_pool`). Side-effect-free so
// they unit-test without a live League client.
//
// ⚠ The LCU match-history endpoint is NOT part of the official public Riot API;
// its shape may drift between client versions. Parsers never throw on unexpected
// shapes — unknown entries are skipped, never fabricated.

export interface ParsedLcuMatch {
  match_id: string; // "LCU_{gameId}" — Riot Match-V5 id'leriyle çakışmasın
  champion_id: number;
  position: string | null;
  win: boolean;
  kills: number;
  deaths: number;
  assists: number;
  duration_secs: number;
  queue_id: number;
  played_at: number; // unix saniye
  cs: number; // lane minions + neutral monsters
  // LCU match history timeline taşımaz → cs_at_10/deaths_pre_14 yok; yalnız
  // visionScore stats'ta mevcut (yoksa dürüst null).
  vision_score: number | null;
}

export interface ParsedLcuMastery {
  champion_id: number;
  level: number;
  points: number;
  last_play_time: number | null;
}

type Json = Record<string, unknown>;

const asObj = (v: unknown): Json => (v && typeof v === "object" ? (v as Json) : {});
const asArr = (v: unknown): unknown[] => (Array.isArray(v) ? v : []);
const num = (v: unknown, fallback = 0): number =>
  typeof v === "number" && Number.isFinite(v) ? v : fallback;

/** puuid'nin participantId'si (participantIdentities üzerinden). */
function localParticipantId(game: Json, puuid: string): number | null {
  for (const id of asArr(game.participantIdentities)) {
    const entry = asObj(id);
    const player = asObj(entry.player);
    if (player.puuid === puuid) {
      const pid = entry.participantId;
      return typeof pid === "number" ? pid : null;
    }
  }
  return null;
}

function parseSingleGame(gameRaw: unknown, puuid: string): ParsedLcuMatch | null {
  const game = asObj(gameRaw);
  const gameId = game.gameId;
  if (typeof gameId !== "number") return null;

  const participantId = localParticipantId(game, puuid);
  if (participantId === null) return null;

  const participant = asArr(game.participants)
    .map(asObj)
    .find((p) => p.participantId === participantId);
  if (!participant) return null;

  const championId = participant.championId;
  if (typeof championId !== "number") return null;

  const stats = asObj(participant.stats);
  const lane = asObj(participant.timeline).lane;
  const position =
    typeof lane === "string" && lane !== "" && lane !== "NONE" ? lane : null;

  return {
    match_id: `LCU_${gameId}`,
    champion_id: championId,
    position,
    win: stats.win === true,
    kills: num(stats.kills),
    deaths: num(stats.deaths),
    assists: num(stats.assists),
    duration_secs: num(game.gameDuration),
    queue_id: num(game.queueId),
    played_at: Math.floor(num(game.gameCreation) / 1000),
    cs: num(stats.totalMinionsKilled) + num(stats.neutralMinionsKilled),
    vision_score: typeof stats.visionScore === "number" ? stats.visionScore : null,
  };
}

/** `/lol-match-history/v1/products/lol/{puuid}/matches` → yalnız puuid'nin maçları. */
export function parseLcuMatchHistory(
  json: unknown,
  puuid: string,
): ParsedLcuMatch[] {
  const games = asArr(asObj(asObj(json).games).games);
  const result: ParsedLcuMatch[] = [];
  for (const game of games) {
    const parsed = parseSingleGame(game, puuid);
    if (parsed) result.push(parsed);
  }
  return result;
}

/** `/lol-champion-mastery/v1/local-player/champion-mastery` → tüm geçerli girdiler. */
export function parseLcuMastery(json: unknown): ParsedLcuMastery[] {
  if (!Array.isArray(json)) return [];
  const result: ParsedLcuMastery[] = [];
  for (const entry of json.map(asObj)) {
    const championId = entry.championId;
    if (typeof championId !== "number") continue;
    result.push({
      champion_id: championId,
      level: num(entry.championLevel),
      points: num(entry.championPoints),
      last_play_time:
        typeof entry.lastPlayTime === "number" ? entry.lastPlayTime : null,
    });
  }
  return result;
}

/**
 * current-summoner yanıtından (game_name, tag_line). Öncelik:
 * gameName+tagLine → displayName "Name#TAG" → displayName olduğu gibi.
 */
export function parseLcuSummonerName(summoner: unknown): {
  gameName: string;
  tagLine: string;
} {
  const s = asObj(summoner);
  const gameName = typeof s.gameName === "string" ? s.gameName.trim() : "";
  const tagLine = typeof s.tagLine === "string" ? s.tagLine.trim() : "";
  if (gameName) return { gameName, tagLine };

  const display =
    typeof s.displayName === "string" && s.displayName.trim()
      ? s.displayName.trim()
      : "Summoner";
  const idx = display.lastIndexOf("#");
  if (idx > 0) {
    return { gameName: display.slice(0, idx), tagLine: display.slice(idx + 1) };
  }
  return { gameName: display, tagLine: "" };
}

/**
 * LCU match-history → `(champion_id, count)` çiftleri, çok oynanandan aza
 * (champ_pool.rs `compute_champion_pool` paritesi). Eşleşmeyen puuid'li
 * oyunlar sessizce atlanır; championId 0 sayılmaz.
 */
export function computeChampionPool(
  historyJson: unknown,
  puuid: string,
): [number, number][] {
  const counts = new Map<number, number>();
  for (const gameRaw of asArr(asObj(asObj(historyJson).games).games)) {
    const game = asObj(gameRaw);
    const participantId = localParticipantId(game, puuid);
    if (participantId === null) continue;
    const participant = asArr(game.participants)
      .map(asObj)
      .find((p) => p.participantId === participantId);
    const championId = participant?.championId;
    if (typeof championId === "number" && championId !== 0) {
      counts.set(championId, (counts.get(championId) ?? 0) + 1);
    }
  }
  return [...counts.entries()].sort((a, b) => b[1] - a[1]);
}
